src/Morfologik.Stemming/DictionaryMetadata.cs

```diff
@@ -98,7 +99,7 @@ public sealed class DictionaryMetadata
         /// <summary>
         /// Gets all metadata attributes.
         /// </summary>
-        public IDictionary<DictionaryAttribute, string> Attributes => attributes.AsReadOnly();
+        public IDictionary<DictionaryAttribute, string> Attributes => JCG.Extensions.DictionaryExtensions.AsReadOnly(attributes);
```

Comment:
Should this be of type `IReadOnlyDictionary<DictionaryAttribute, string>`? Would be a breaking change, so could be a separate PR/issue.

Reply:
Unfortunately, while all BCL types that implement `IDictionary<TKey, TValue>` also implement `IReadOnlyDictionary<TKey, TValue>` there is no guarantee this will always be the case. So, casting is not guaranteed to work, and we need `IDictionary<TKey, TValue>` to call other APIs. Basically, this is an all or nothing proposition. We would need to change all APIs to be `IReadOnlyDictionary<TKey, TValue>`.

Also, this component is already released and there has only been 1 patch released higher than 2.1.7. So, we would need to introduce breaking changes in a patch. Not saying that we shouldn't, but it wouldn't be very nice if we had breaking changes (believe it or not, there are other libraries depending on this other than Lucene.NET, such as https://www.nuget.org/packages/Clara.Analysis.Morfologik#dependencies-body-tab).

> NOTE: I just noticed that there are 3 modules of this that haven't been ported [`morfologik-fsa-builders`](https://github.com/morfologik/morfologik-stemming/tree/master/morfologik-fsa-builders), [`morfologik-speller`](https://github.com/morfologik/morfologik-stemming/tree/master/morfologik-speller), and [`morfologik-tools`](https://github.com/morfologik/morfologik-stemming/tree/master/morfologik-tools). I don't recall whether there was a reason for that other than the fact that Lucene.NET doesn't directly depend on them.

Reply:
Good point re: breaking change, I think that could be split out as an issue for future improvement.

Reply:
Well, in this case we could do it because there are specific closing types so we can see where in Lucene.NET this is valid to do. But in general, we have been avoiding read only collection interfaces for compatibility reasons. We cannot always make assumptions about where the Lucene designers intended the output of one member to be used in the input of another.

Reply:
I opened #6, #7, #8, and #9 to deal with these.

Reply:
I opened #7, #8, and #9 to deal with these.